File: Week3/code/Vectorize2.R
# Runs the stochastic Ricker equation with gaussian fluctuations

rm(list=ls())

stochrick<-function(p0=runif(1000,.5,1.5),r=1.2,K=1,sigma=0.2,numyears=100)
{
  #initialize
  N<-matrix(NA,numyears,length(p0))
  N[1,]<-p0
  
  for (pop in 1:length(p0)){#loop through the populations
    
    for (yr in 2:numyears){ #for each pop, loop through the years

      N[yr,pop] <- N[yr-1,pop] * exp(r * (1 - N[yr - 1,pop] / K) + rnorm(1,0,sigma))
    
    }
  
  }
 return(N)

}


stochrickvect<-function(p0=runif(1000,.5,1.5),r=1.2,K=1,sigma=0.2,numyears=100){
#sets p0 to be a population of 1000 individuals of random identity between 0.5 and 1.5
#sets the intrinsic growth rate (r) to be 1.2
#sets the carrying capacity of the population (k) to be 1
#sets the standard deviation for generating stochacity (sigma) to be 0.2
#sets the running period to be 100 iterations

    N<-matrix(NA,numyears,length(p0)) #creates a matrix of NA, numyears (100), and number of arguments in p0 (100)
    N[1,]<-p0 #sets the first column to be the values of p0 (1000 randomly generated numbers between 0.5 and 1.5)
    for (yr in numyears){
        N[yr,] <- N[yr-1,] * exp(r * (1 - N[yr - 1,] / K) + rnorm(1,0,sigma)) #Runs stochastic Ricker model with Gaussian Fluctuations (rnorm...)
    }

    return(N)}

print("Loopy Stochastic Ricker takes:")
print(system.time(res2<-stochrick()))

print("Vectorized Stochastic Ricker takes:") 
print(system.time(res2<-stochrickvect()))

File: Week3/code/Vectorize2.py
#!/usr/bin/env python3

""" Runs the stochastic Ricker equation with gaussian fluctuations """

__appname__ = '[Vectorize2.py]'
__author__ = 'Victoria Blanchard (vlb19@ic.ac.uk)'
__version__ = '0.0.1'
__license__ = "License for this code/program"

#####################################################
# Imports
#####################################################

import numpy as np
import random
import time

#####################################################
# Runs loopy stochastic Ricker model
#####################################################

def stochrick():
    # generate a list of random numbers between 0.5 and 1.5 1000 times
    p0 = []
    for i in range(0,1000):
        nextvalue = random.uniform(0.5,1.5)
        p0.append(nextvalue)
    # define other variables in stochastic ricker model
    r = 1.2 
    K = 1
    sigma = 0.2
    numyears = 100

    # generate a matrix of zeros with numyears number of rows and len(p0) number of columns
    N = np.zeros((numyears,len(p0))) 
    N[0] = p0

    # loop through each row in the matrix
    for pop in range (0, len(p0)): 
        # loop through column in the matrix
        for yr in range(1, numyears):
            # Run the equation for the stochastic Ricker Model
            N[yr,pop] = N[yr-1,pop]* np.exp(r* (1 - N[yr - 1, pop] / K) + np.random.normal(0.5,sigma))

    return(N)

##################################################
### Vectorised stochastic Ricker Model
##################################################

def stochrickvect():
    p0 = []
    for i in range(0,1000):
        nextvalue = random.uniform(0.5,1.5)
        p0.append(nextvalue)

    r = 1.2 
    K = 1
    sigma = 0.2
    numyears = 100
    N = np.zeros((numyears,len(p0))) 
    N[0] = p0
    for yr in range (1,numyears): # each column is a vector now, apply the function to each vector not each element = faster
        N[yr,] = N[yr-1,]* np.exp(r* (1 - N[yr - 1,] / K) + np.random.normal(0.5,sigma))
    return(N)


start_time = time.time() #store the current time as a variable
stochrick() #run the function
runningtime = time.time() - start_time #running time = time now - time before running the script
print("Loopy Stochastic Ricker takes:", runningtime)

start_time = time.time()
stochrickvect()
runningtime = time.time() - start_time
print("Vectorized Stochastic Ricker takes:", runningtime)

